File: testdata/fixtures/go/vulnerable/open_redirect.go
package vulnerable

import (
	"net/http"
)

// VULN: Open redirect - user controls the redirect destination URL.
// Should trigger taint analysis for go.http.redirect sink with user input source.

func HandleRedirect(w http.ResponseWriter, r *http.Request) {
	target := r.URL.Query().Get("url")

	http.Redirect(w, r, target, http.StatusFound)
}

func HandleLoginRedirect(w http.ResponseWriter, r *http.Request) {
	returnTo := r.URL.Query().Get("return_to")

	// Authenticate user ...
	// Then redirect to the return URL without validation
	http.Redirect(w, r, returnTo, 302)
}

File: testdata/fixtures/javascript/vulnerable/open_redirect.ts
import { Request, Response } from 'express';

// Juice Shop style: open redirect with user-controlled URL
export function handleRedirect(req: Request, res: Response): void {
  const redirectTo = req.query.to as string;

  // VULNERABLE: user-controlled redirect target, no URL validation
  // Attacker can use: /redirect?to=https://evil.com
  res.redirect(redirectTo);
}

export function handleLogin(req: Request, res: Response): void {
  const { username, password } = req.body;
  const returnUrl = req.query.returnUrl as string;

  // Simulate authentication
  if (username === 'admin') {
    // VULNERABLE: open redirect after login
    // location.href assignment with user-controlled url
    res.send(`<script>window.location.href = '${returnUrl}';</script>`);
  } else {
    res.status(401).send('Unauthorized');
  }
}
